sdk/fuzz_test.go

```diff
@@ -0,0 +1,150 @@
+		"5cHRpb25JbmZvcm1hdGlvbiI6eyJ0eXBlIjoic3BsaXQiLCJwb2xpY3kiOiJleUoxZFdsa0" +
+		"lqb2lZakF3TW1WaU9USXROV0l4TkMweE1XVm1MVGt4TW1NdFlXRTFZalprWlRjMVlUQmpJa" +
+		"XdpWW05a2VTSTZleUprWVhSaFFYUjBjbWx5ZFhSbGN5STZiblZzYkN3aVpHbHpjMlZ0SWpw" +
+		"dWRXeHNmWDA9Iiwia2V5QWNjZXNzIjpbeyJ0eXBlIjoid3JhcHBlZCIsInVybCI6ImV4YW1" +
+		"wbGUuY29tIiwicHJvdG9jb2wiOiJrYXMiLCJ3cmFwcGVkS2V5IjoiV1dZait3anNMQmtrU2" +
+		"FjTzZ2dEpJaTBLMUJQMVhtT2lzcFNrdm8wRm5QV0ZLM050UTVzN3YwOVpqQ05NV0JRK1VPa" +
+		"VhUTVNWa1JkNUdsTHlMblg3bjY4dDBmSDk0RnMyTnRjcFJwMSt6YStjdzVGRldFQy9uQUJp" +
+		"TmtPdldLeHdqeG5YQ1pEazZ4U3o1ZHdCT1MraUVCYXJ6WGMzR3oxR2JYcm5Ka0YvaitUUDR" +
+		"rbTJUYUpXN0cybFJaQ0J6T1M5RkpoSEFIcFBIcFF4V2tNK2FuZjJ1WExRV1UxT00vaHFVRz" +
+		"VFUG9nR0pYM3MxaVRmek4xNFhiczU5TmYyOU1rc284VjhJSnNOWVRPblBIejY4Q3VvOGdjc" +
+		"XZHd3J0a3FKQmlmYVM3N1FRQWxwUTcrSU9GME9ZSjh1WTZLZG1najltSU1aRUVaYkI3V2hO" +
+		"blNBbG9paWZBPT0iLCJwb2xpY3lCaW5kaW5nIjp7ImFsZyI6IkhTMjU2IiwiaGFzaCI6Ilp" +
+		"UY3pZMkV5WkdReVkySTJNRGN4WmpnellXVTVNRGsxWXpnNU5XWXhOalUwWVRjNE5tTXpPV1" +
+		"EwTW1JM05qQmxOemxsTmpWaVltWTRZalUyWkdNd013PT0ifX1dLCJtZXRob2QiOnsiYWxnb" +
+		"3JpdGhtIjoiQUVTLTI1Ni1HQ00iLCJpdiI6IiIsImlzU3RyZWFtYWJsZSI6dHJ1ZX0sImlu" +
+		"dGVncml0eUluZm9ybWF0aW9uIjp7InJvb3RTaWduYXR1cmUiOnsiYWxnIjoiSFMyNTYiLCJ" +
+		"zaWciOiJNRFZqTURReE1EWmtNR00wWlRRMllUZG1PRFJrWVRJM09UZGlPREk1WVRWak5EVX" +
+		"hPRGs0TkRreE1HWTFaV1kxTXpKbVpHWmtZMlkwWWprek0yVmhOZz09In0sInNlZ21lbnRIY" +
+		"XNoQWxnIjoiR01BQyIsInNlZ21lbnRTaXplRGVmYXVsdCI6MjA5NzE1MiwiZW5jcnlwdGVk" +
+		"U2VnbWVudFNpemVEZWZhdWx0IjoyMDk3MTgwLCJzZWdtZW50cyI6W3siaGFzaCI6IlpETm1" +
+		"OVFkyWW1FM05XWmxZVGt4TWpNeU5ESXdPRFZoWWpGbVpUbGhNVEU9Iiwic2VnbWVudFNpem" +
+		"UiOjMsImVuY3J5cHRlZFNlZ21lbnRTaXplIjozMX1dfX0sInBheWxvYWQiOnsidHlwZSI6I" +
+		"nJlZmVyZW5jZSIsInVybCI6IjAucGF5bG9hZCIsInByb3RvY29sIjoiemlwIiwibWltZVR5" +
+		"cGUiOiJhcHBsaWNhdGlvbi9vY3RldC1zdHJlYW0iLCJpc0VuY3J5cHRlZCI6dHJ1ZX19UEs" +
+		"HCALoriwCBQAAAgUAAFBLAQItAC0ACAAAAH11LzEke7o5HwAAAB8AAAAJAAAAAAAAAAAAAA" +
+		"AAAAAAAAAwLnBheWxvYWRQSwECLQAtAAgAAAB9dS8xAuiuLAIE///tBQAADwAAAAAAAAAAA" +
+		"AAAAABWAAAAMC5tYW5pZmVzdC5qc29uUEsFBgAAAAACAAIAdAAAAJUFAAAAAA=="))
+
+	f.Fuzz(func(t *testing.T, data []byte) {
+		r, err := sdk.LoadTDF(bytes.NewReader(data))
+		if err != nil {
+			assert.Nil(t, r)
+			return
+		}
+		assert.NotNil(t, r)
+		// TODO fuzz r somewhat
+	})
+}
+
+func FuzzReadPolicyBody(f *testing.F) {
+	pb := &PolicyBody{
+		mode: 0,
+		rp: remotePolicy{
+			url: ResourceLocator{
+				protocol: 0,
+				body:     "example.com",
+			},
+		},
+	}
+	f.Add(writeBytes(pb.writePolicyBody))
+	pb = &PolicyBody{
+		mode: 1,
+		ep: embeddedPolicy{
+			lengthBody: 3,
+			body:       []byte("foo"),
+		},
+	}
+	f.Add(writeBytes(pb.writePolicyBody))
+
+	f.Fuzz(func(t *testing.T, data []byte) {
+		pb = &PolicyBody{}
+		err := pb.readPolicyBody(bytes.NewReader(data))
+		if err != nil {
+			assert.Zerof(t, *pb, "unexpected %v", *pb)
+			return
+		}
+		switch pb.mode {
```

Comment:
I am somewhat cautious to put this type of validation into a fuzz test.  If we add a new mode it may not be obvious this test needs to be updated, in part because the test likely wont fail immediately (needing a random mode that now has been added).

In fuzzing we also want to keep validation somewhat minimal so that fuzz iterations can be as fast as possible.  Usually I defer to only the default validation which is:
* It does not panic
* It returns quickly (I believe 10 seconds is the default)
* There is no out of memory condition hit (unfortunately this is system dependent, so you may want to consider adding additional process limits if on a system with a lot of ram)

Instead I think this type of validation is better as a unit test, where an invalid mode is used and validated that the `undefined policy mode` error is returned.

Reply:
fair enough. It is interesting that when not run with `-fuzz`, the default behavior is to run the tests on the seeds - basically turning the fuzz test into a table driven test, in golang parlance. For more complex 'expected error' tests, it may make sense to share the seeds from the table driven tests that exercise error conditions explicitly